### PR TITLE
[Hardware][Intel] OpenVINO vLLM backend

### PR DESCRIPTION
**Pull request by ilya-lavrenov**
Adds OpenVINO vLLM backend, described here https://github.com/vllm-project/vllm/issues/5377

Below and performance measurements:
- VLLM - vLLM CPU backend is taken from commit https://github.com/vllm-project/vllm/commit/388596c91437a51d428a447594e9faec340c29b2
- OV (-Q) - OpenVINO vLLM backend w/o KV cache / weights compression, default vLLM scheduling mode
- OV (+Q) - OpenVINO vLLM backend w/ KV cache / weights compression, chunked prefill scheduling

<details>
  <summary>OpenVINO vLLM vs vLLM CPU TTFT</summary>

image

</details>

<details>
  <summary>OpenVINO vLLM vs vLLM CPU TPOT</summary>

image

</details>

### CODE REVIEW

File: vllm/utils.py
Comment (robertgshaw2-neuralmagic) [inline vllm/utils.py:34]: nit: this is covered by `half`?

File: vllm/utils.py
Comment (robertgshaw2-neuralmagic) [inline vllm/utils.py:39]: uint8 or int8 please

File: vllm/worker/openvino_model_runner.py
Comment (sroy745) [inline vllm/worker/openvino_model_runner.py:38]: Can you add some tests for the ModelRunner?

File: vllm/worker/openvino_model_runner.py
Comment (sroy745) [inline vllm/worker/openvino_model_runner.py:223]: where is multi_modal_input_list getting populated? It seems like you are defining it in L114 but never setting it?

**Review by sroy745 — COMMENTED** [review]
Is there any plan for running an e2e test for this in openvino compatible hardware?

File: docs/source/getting_started/openvino-installation.rst
Comment (l-bat) [inline docs/source/getting_started/openvino-installation.rst:60]:
```suggestion
    $ PIP_PRE=1 PIP_EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu https://storage.openvinotoolkit.org/simple/wheels/nightly/" VLLM_TARGET_DEVICE=openvino python -m pip install -v .
```

File: vllm/worker/openvino_worker.py
Comment (ilya-lavrenov) [reply vllm/worker/openvino_worker.py:63]: > Wondering if it is possible to move the logic of updating the head_size into the ModelConfig itself and update the value of ModelConfig.head_size?

as I see `ModelConfig` does not have such field `head_size`, which can be changed.
It's a function which can return its value depending on several conditions:
```python
    def get_head_size(self) -> int:
        if hasattr(self.hf_text_config, "head_dim"):
            return self.hf_text_config.head_dim
        # FIXME(woosuk): This may not be true for all models.
        return (self.hf_text_config.hidden_size //
                self.hf_text_config.num_attention_heads)
```

I have updated the code to use `self.head_size` to reduce confusion.

File: vllm/worker/openvino_worker.py
Comment (ilya-lavrenov) [reply vllm/worker/openvino_worker.py:127]: It has similar logic, but different implementation details:
- Uses `OpenVINOCacheEngine`
- Tensors to store KV caches are `ov.Tensor`s. In future PRs, when we will add GPU support, it will be `ov.RemoteTensor` to keep GPU dedicated memory.
- It mentions OpenVINO specific env vars like `VLLM_OPENVINO_KVCACHE_SPACE`
- It uses `OpenVINOModelRunner`

File: vllm/executor/openvino_executor.py
Comment (ilya-lavrenov) [reply vllm/executor/openvino_executor.py:112]: I've just used similar scheme as for CPU backend:
https://github.com/vllm-project/vllm/blob/dd248f76756adba4a1637b882e79ab639f957feb/vllm/executor/cpu_executor.py#L114-L137

> ModelConfig (vllm/config.py)

I supposed it's core vLLM entity which is not touched by backend specific things, is not it?
I see that `_get_and_verify_dtype` from `vllm/config.py` is not parametrized with `DeviceConfig`, assuming no backend specific code.

File: vllm/executor/openvino_executor.py
Comment (ilya-lavrenov) [reply vllm/executor/openvino_executor.py:125]: The same answer as for `_verify_and_get_model_config`

I don't mind if we could integrate our logic there, but `vllm/config.py`  does not assume backend specific code.
Maybe it's a good point for future improvements.

File: vllm/attention/backends/openvino.py
Comment (ilya-lavrenov) [reply vllm/attention/backends/openvino.py:64]: good note, added more comprehensive description

File: vllm/worker/openvino_model_runner.py
Comment (ilya-lavrenov) [reply vllm/worker/openvino_model_runner.py:38]: Can we implement as next step in our RFC for OpenVINO integration?

BTW, I don't see any tests for any model runners except CUDA ones. How it supposed to run such tests? Does vLLM provide an ability to build vLLM with specific backend on CI service and run backend tests? So far, I have seen small scripts like `run-cpu-test.sh`, which can ensure e2e flow works and integration is not broken.

File: vllm/attention/selector.py
Comment (ilya-lavrenov) [reply vllm/attention/selector.py:68]: added, thank you for this hint.

File: vllm/worker/openvino_model_runner.py
Comment (ilya-lavrenov) [reply vllm/worker/openvino_model_runner.py:38]: BTW, I don't see that scripts `run-xpu-test.sh`, `run-cpu-test.sh` are currently used, while previously they were mentioned in some CI scripts. How can we ensure e2e on simple model is run on CI?

File: vllm/utils.py
Comment (ilya-lavrenov) [reply vllm/utils.py]: dropped

File: vllm/utils.py
Comment (ilya-lavrenov) [reply vllm/utils.py]: dropped

File: vllm/engine/async_llm_engine.py
Comment (ilya-lavrenov) [reply vllm/engine/async_llm_engine.py]: good point, thank you! Fixed per your suggestion

File: vllm/executor/openvino_executor.py
Comment (ilya-lavrenov) [reply vllm/executor/openvino_executor.py]: done, thank you!

File: vllm/executor/openvino_executor.py
Comment (ilya-lavrenov) [reply vllm/executor/openvino_executor.py]: done, thank you!

File: vllm/model_executor/model_loader/openvino.py
Comment (ilya-lavrenov) [reply vllm/model_executor/model_loader/openvino.py]: dropped this env var, because it more useful for debugging rather than for real deployment scenarios.

File: vllm/model_executor/model_loader/openvino.py
Comment (ilya-lavrenov) [reply vllm/model_executor/model_loader/openvino.py]: good point, dropped this prefix!

File: vllm/model_executor/model_loader/openvino.py
Comment (ilya-lavrenov) [reply vllm/model_executor/model_loader/openvino.py]: dropped this env var, because it more useful for debugging rather than for real deployment scenarios.

File: vllm/model_executor/model_loader/openvino.py
Comment (ilya-lavrenov) [reply vllm/model_executor/model_loader/openvino.py]: done, thank you!

File: vllm/worker/openvino_model_runner.py
Comment (ilya-lavrenov) [reply vllm/worker/openvino_model_runner.py]: dropped this variable as OpenVINO backend does not cover multi modal models. 
It's a good direction for future improvement of this integration as we support it w/o serving. Hope, we can safely enable such models in future PRs after validation on our side.

File: docs/source/getting_started/openvino-installation.rst
Comment (ilya-lavrenov) [reply docs/source/getting_started/openvino-installation.rst]: thank you, applied!

**Comment by ilya-lavrenov** [comment]
> Is there any plan for running an e2e test for this in openvino compatible hardware?

Currently, OpenVINO backend supports any CPU, so adding such HW should not be a problem.
How should we proceed with adding our HW? is there any document how backends can organize its testing?

File: vllm/envs.py
Comment (mgoin) [inline vllm/envs.py:221]: Can you note that u8 is allowed as well?

File: vllm/envs.py
Comment (mgoin) [inline vllm/envs.py:225]: This seems to be for model weights, not for kv cache

File: vllm/engine/arg_utils.py
Comment (mgoin) [inline vllm/engine/arg_utils.py:514]: Why is yapf disabled here? Did you have issues with `./format.sh`?

File: vllm/worker/openvino_model_runner.py
Comment (mgoin) [reply vllm/worker/openvino_model_runner.py:38]: This makes sense for a followup PR. Maybe @simon-mo can clarify how to go about new CI backends?

File: docs/source/getting_started/openvino-installation.rst
Comment (ilya-lavrenov) [reply docs/source/getting_started/openvino-installation.rst:6]: > all LLM models

LLaVa is VLM, while by all LLMs I assumes all text LLMs w/o multi model inputs.
Do you think this statement is not fully precise?

MoE is also supported.

File: vllm/envs.py
Comment (ilya-lavrenov) [reply vllm/envs.py:221]: good point, added description for u8 as well

File: vllm/envs.py
Comment (ilya-lavrenov) [reply vllm/envs.py]: thank you for noticing it, updated this copy-paste comment

File: vllm/engine/arg_utils.py
Comment (ilya-lavrenov) [reply vllm/engine/arg_utils.py]: I have not used it 😄 

With such script formatting is really simple, thank you for pointing out.

File: docs/source/getting_started/openvino-installation.rst
Comment (mgoin) [reply docs/source/getting_started/openvino-installation.rst]: If MoEs are supported I think that is sufficient complexity :) thanks for clarifying!

File: vllm/worker/openvino_worker.py
Comment (mgoin) [inline vllm/worker/openvino_worker.py:344]: Is this needed given we currently allow a world size of 1?

**Review by mgoin — APPROVED** [review]
Appreciate the clarity and improvements. I think this is a great start for this backend and looking forward to automated CI, performance, and functionality improvements!

**Review by sroy745 — COMMENTED** [review]
Thanks for addressing the comments! LGTM.

File: vllm/worker/openvino_worker.py
Comment (ilya-lavrenov) [reply vllm/worker/openvino_worker.py:344]: I've copied this code as is from CPU worker and supposed that it's harmless even for world size 1, right?

**Comment by ilya-lavrenov** [comment]
Thank you @mgoin and @sroy745 
Please, let me know if I can improve something in current PR!
Hope, we can merge it soon to avoid endless merge conflicts 😃

File: vllm/worker/openvino_worker.py
Comment (mgoin) [reply vllm/worker/openvino_worker.py:344]: Yeah it is harmless. Just checking it is supposed to be disabled.

**Comment by mgoin** [comment]
The CI is a bit messed up at the moment, so if it fails again I will let you know when it is in a good state to rebase for green checks.

**Comment by WoosukKwon** [comment]
Hi @ilya-lavrenov, thanks for submitting the PR.

We actually have some concerns about the maintenance cost of the OpenVINO backend. Since it isn't based on PyTorch, it can easily break whenever core changes are made to vLLM. Additionally, as most of our committers and reviewers aren't familiar with OpenVINO, it is difficult for us to work on the OpenVINO backend.

For these reasons, we hope your team can take responsibility for testing and maintaining the OpenVINO backend. When core changes are made in vLLM, we won't require the PR committers to test and fix the OpenVINO backend. Instead, we will expect your team to handle this after the PR is merged. Hopefully, after ~6 months, vLLM will be more stable with more mature abstractions for different backends, making maintenance easier. However, at this early stage, there may be significant changes in the system architecture, and we hope your team can address any issues with the OpenVINO backend.

Please let us know if you have any questions. We'd be happy to discuss this further!

**Comment by ilya-lavrenov** [comment]
> Hi @ilya-lavrenov, thanks for submitting the PR.
> 
> We actually have some concerns about the maintenance cost of the OpenVINO backend. Since it isn't based on PyTorch, it can easily break whenever core changes are made to vLLM. Additionally, as most of our committers and reviewers aren't familiar with OpenVINO, it is difficult for us to work on the OpenVINO backend.
> 
> For these reasons, we hope your team can take responsibility for testing and maintaining the OpenVINO backend. When core changes are made in vLLM, we won't require the PR committers to test and fix the OpenVINO backend. Instead, we will expect your team to handle this after the PR is merged. Hopefully, after ~6 months, vLLM will be more stable with more mature abstractions for different backends, making maintenance easier. However, at this early stage, there may be significant changes in the system architecture, and we hope your team can address any issues with the OpenVINO backend.
> 
> Please let us know if you have any questions. We'd be happy to discuss this further!

Hi @WoosukKwon!
Sure, we will be responsible for OpenVINO integration and check its consistency with other backends.

Please, mention us explicitly in such PRs where common logic can be changed.

**Comment by JNewman-cell** [comment]
Setting the environment variable ENGINE_ITERATION_TIMEOUT_S to >60 increases the async io timeout which may be causing the issue because inference on CPU is very slow.

**Pull request merged**
